hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenClaimAirdropHandler.java

```diff
@@ -36,25 +71,169 @@
+        this.validator = validator;
     }
 
     @Override
-    public void preHandle(@NonNull PreHandleContext context) throws PreCheckException {}
+    public void preHandle(@NonNull PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        final var op = context.body().tokenClaimAirdrop();
+        requireNonNull(op);
+        final var pendingAirdrops = op.pendingAirdrops();
+
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+
+        for (final var pendingAirdrop : pendingAirdrops) {
+            AccountID receiverId = pendingAirdrop.receiverIdOrThrow();
+            Account account = accountStore.getAccountById(receiverId);
+            requireNonNull(account);
+            Key key = account.key();
+            validateTruePreCheck(key != null, KEY_NOT_PROVIDED);
+            context.requireKey(key);
+        }
+    }
 
     @Override
-    public void pureChecks(@NonNull TransactionBody txn) throws PreCheckException {}
+    public void pureChecks(@NonNull TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn);
+
+        final var op = txn.tokenClaimAirdrop();
+        requireNonNull(op);
+
+        final List<PendingAirdropId> pendingAirdrops = op.pendingAirdrops();
+        if (pendingAirdrops.isEmpty()) {
+            throw new PreCheckException(EMPTY_PENDING_AIRDROP_ID_LIST);
+        }
+
+        final int numAirdrops = pendingAirdrops.size();
+        final Set<PendingAirdropId> uniqueAirdrops = Set.copyOf(pendingAirdrops);
+        if (numAirdrops != uniqueAirdrops.size()) {
+            throw new PreCheckException(PENDING_NFT_AIRDROP_ALREADY_EXISTS);
+        }
+    }
 
     @Override
-    public void handle(@NonNull HandleContext context) throws HandleException {}
+    public void handle(@NonNull HandleContext context) throws HandleException {
+        final var op = context.body().tokenClaimAirdropOrThrow();
+        final var tokensConfig = context.configuration().getConfigData(TokensConfig.class);
+        validateTrue(
+                op.pendingAirdrops().size() < tokensConfig.maxAllowedPendingAirdropsToClaim(),
+                MAX_PENDING_AIRDROP_ID_EXCEEDED);
+
+        final var pendingAirdropStore = context.storeFactory().writableStore(WritableAirdropStore.class);
+        final var accountStore = context.storeFactory().writableStore(WritableAccountStore.class);
+        final var tokenStore = context.storeFactory().readableStore(ReadableTokenStore.class);
+        final var tokenRelStore = context.storeFactory().writableStore(WritableTokenRelationStore.class);
+        final var recordBuilder = context.savepointStack().getBaseBuilder(CryptoTransferStreamBuilder.class);
+
+        final List<TokenTransferList> transfers = new ArrayList<>();
+        final List<Token> tokensToAssociate = new ArrayList<>();
+        final AccountID receiverId = op.pendingAirdrops().getFirst().receiverId();
```

Comment:
The receiver is different for each airdrop right ? Why is this outside for loop 🤔

Reply:
Added [here](https://github.com/hashgraph/hedera-services/pull/14738/files#diff-01f69a88b2ddfcb52d19ed653d2acd23144a0226d7c728a0d4424818a5f9b4eaR105-R109). But I'm not sure about the response.